src/templates/expressions.ts

```diff
@@ -125,58 +137,63 @@ export class ExpressionMeta {
   }
 }
 
-export class Expression {
+export abstract class Expression {
   module = 'expressions';
   type = 'Expression';
   meta: ExpressionMeta;
-  segments: string[];
+  segments: Segments;
 
-  constructor(meta) {
+  constructor(meta: ExpressionMeta) {
     this.meta = meta;
   }
 
-  serialize() {
+  serialize(): string {
     return serializeObject.instance(this, this.meta);
   }
 
-  toString() {
+  toString(): string {
     return this.meta && this.meta.source;
   }
 
-  truthy(context) {
+  truthy(context: Context) {
     const blockType = this.meta.blockType;
     if (blockType === 'else') return true;
     const value = this.get(context, true);
     const truthy = templateTruthy(value);
     return (blockType === 'unless') ? !truthy : truthy;
   }
 
-  get(_context, _flag?): any { }
+  abstract get(_context: Context, _flag?: boolean): any;
 
   // Return the expression's segment list with context objects
-  resolve(_context): any { }
+  resolve(_context: Context): Segments | undefined { return undefined; }
 
   // Return a list of segment lists or null
-  dependencies(_context, _options): any { }
+  dependencies(_context: Context, _options: any): string[][] | undefined { return undefined; }
 
   // Return the pathSegments that the expression currently resolves to or null
-  pathSegments(context) {
+  pathSegments(context: Context): Segments | null {
     const segments = this.resolve(context);
     return segments && pathSegments(segments);
   }
 
-  set(context, value) {
+  set(context: Context, value: Value): void {
     const segments = this.pathSegments(context);
     if (!segments) throw new Error('Expression does not support setting');
     context.controller.model._set(segments, value);
   }
 
-  _resolvePatch(context, segments) {
-    return (context && context.expression === this && context.item != null) ?
-      segments.concat(context) : segments;
+  _resolvePatch(context: Context, segments: Segments): Segments {
+    if (context.item == null) {
```

Comment:
Let's either add a safety check for null/undefined `context`, or verify that it'll never be nullish in practice in a real app or tests.